core/templates/services/context.service.ts

```diff
@@ -189,7 +192,7 @@ export class ContextService {
   }
 
   // Add constants for entity type.
-  getEntityType(): string {
+  getEntityType(): string | undefined {
```

Comment:
What's the valid use case for undefined? What does it mean when you return undefined. Wherever this function is used, are we checking for undefined?

Reply:
I came to this conclusion of making the return type to be string | undefined after some testing, `getEntityType` returns some string like 'collection',  'exploration', 'topic',  'skill',  'story',  'question', depending on the particular place it is used. The function returns undefined at every other place apart from places where these aren't used. Suppose this function is indirectly called at some place dealing with subtopic or through some thumbnail-uploader then that causes to stop execution at that particular time throwing that error. I previously tried to throw error but that lead to constant failure of two e2e. So I think this should be made string | undefined. Here's the initial commit https://github.com/oppia/oppia/pull/13178/commits/532fc68a842637a236704d2f751728a1461339cd
e2e failures:
https://github.com/oppia/oppia/runs/2917610311
https://github.com/oppia/oppia/runs/2925276877

Reply:
Thanks for the explanation. This is something that should be escalated and not pacified. Did you run the e2e tests locally to figure out where it was failing (or at what place the errors show up in console)? If with that changes, the e2e tests are not passing, it implies that there could be something wrong with our codebase or we could have misunderstood a use case.

Reply:
It was wrong usage of `getExplorationId` at exploration-footer.component.ts file that was causing an e2e to fail. I have added a check for it to be not undefined. PTAL

Reply:
We can't really get an explorationID for a URL like this `http://localhost:8181/learn/math/dummy-topic-one/practice/session?selected_subtopic_ids=1` required in exploration-footer, so we need to check for it to be not undefined.

Reply:
Ah... thanks for the explanation.